sdk/eventhub/azure-eventhub/azure/eventhub/_pyamqp/link.py

```diff
@@ -241,11 +241,10 @@ def attach(self):
         self._set_state(LinkState.ATTACH_SENT)
 
     def detach(self, close=False, error=None):
-        if self.state in (LinkState.DETACHED, LinkState.ERROR):
+        if self.state in (LinkState.DETACHED, LinkState.DETACH_SENT, LinkState.ERROR):
             return
         try:
-            self._check_if_closed()
```

Comment:
If we had previously received an incoming Detach due to an error scenario, that error would be assigned to `self._error` on line 232, which would then be raised here (though this had been incorrectly convoluted with `self._is_closed`.
By removing this line, where will that error ultimately get propagated? And does this have flow-on effects? 
You mention in your comment above that any errors have already been raised by this point - where?

Reply:
There are 2 places which would be immediate, one at the receiver `[_wait_for_response]`(https://github.com/Azure/azure-sdk-for-python/blob/1f9192a3bc3880edf9b44cc23e1cd3914b8b9b0b/sdk/eventhub/azure-eventhub/azure/eventhub/_pyamqp/receiver.py#L79) , and the other at the sender by calling `[_check_if_closed]`(https://github.com/Azure/azure-sdk-for-python/blob/1f9192a3bc3880edf9b44cc23e1cd3914b8b9b0b/sdk/eventhub/azure-eventhub/azure/eventhub/_pyamqp/sender.py#L164)

I think the above two flows are also slightly different due to the confusion between `is_closed` and detached. The errors would be raised here as they try to perform a send or receive.

Another way errors in `self._error` would be raised is via `link.get_state` which is called in `[client.py]`(https://github.com/Azure/azure-sdk-for-python/blob/1f9192a3bc3880edf9b44cc23e1cd3914b8b9b0b/sdk/eventhub/azure-eventhub/azure/eventhub/_pyamqp/client.py#L550) ( and is part of the bigger while loop of `do_work` and  `client_ready` in the same file )

### Flow impact

`self._error` at the link level only gets populated `_incoming_detach` on line 232. After this the link can be in two states 
Detached or Error (for this case point self._error is populated)

both of these states are in the `if state in return` block of detach. They would never raise there.

if a link is any `ATTACH*` or `DETACH_SENT` state ( current implementation), `self._error` would not have anything. 

Further, in Attach state `is_closed` would be False and would break out from `check_if_closed`.

Detach_sent would be only time `is_closed` is true, but thats what would cause the `Link Error on Detach` to trigger, which is incorrect, we dont have any information from the service yet.  

So currently `check_if_closed` being called here would have an effect for 1 state. 

with the new change, if state is DETACH_SENT, no need to send a detach and at some point a detach would come from the service with any errors.